projects/fal/src/fal/_serialization.py

```diff
@@ -110,92 +79,114 @@ def _pickle_model_field(
+    from threading import Lock
+    from _thread import LockType
 
-    for exception_cls in _get_subclasses(BaseException):
-        dill.pickle(exception_cls, save_exception)
+    def create_lock(locked: bool) -> Lock:
+        lock = Lock()
+        if locked and not lock.acquire(False):
+            raise pickle.UnpicklingError("Cannot acquire lock")
+        return lock
+
+    def pickle_lock(obj: LockType) -> tuple[Callable, tuple]:
+        return create_lock, (obj.locked(),)
+
+    _register(LockType, pickle_lock)
+
+
+def _patch_rlock() -> None:
+    # https://github.com/uqfoundation/dill/blob/70f569b0dd268d2b1e85c0f300951b11f53c5d53/dill/_dill.py#L1317
+    from _thread import RLock as RLockType  # type: ignore[attr-defined]
+
+    def create_rlock(count: int, owner: int) -> RLockType:
+        lock = RLockType()
+        if owner is not None:
+            lock._acquire_restore((count, owner))  # type: ignore[attr-defined]
+        if owner and not lock._is_owned():  # type: ignore[attr-defined]
+            raise pickle.UnpicklingError("Cannot acquire lock")
+        return lock
+
+    def pickle_rlock(obj: RLockType) -> tuple[Callable, tuple]:
+        r = obj.__repr__()
+        count = int(r.split('count=')[1].split()[0].rstrip('>'))
+        owner = int(r.split('owner=')[1].split()[0])
+
+        return create_rlock, (count, owner)
+
+    _register(RLockType, pickle_rlock)
 
 
-@mainify
 def _patch_console_thread_locals() -> None:
-    # NOTE: we __sometimes__ might have to serialize these
     from rich.console import ConsoleThreadLocals
 
-    @dill.register(ConsoleThreadLocals)
-    def save_console_thread_locals(pickler, obj):
-        args = {
-            "theme_stack": obj.theme_stack,
-            "buffer": obj.buffer,
-            "buffer_index": obj.buffer_index,
-        }
+    def create_locals(kwargs: dict) -> ConsoleThreadLocals:
+        return ConsoleThreadLocals(**kwargs)
 
-        def unpickle(kwargs):
-            return ConsoleThreadLocals(**kwargs)
+    def pickle_locals(obj: ConsoleThreadLocals) -> tuple[Callable, tuple]:
+        kwargs = {"theme_stack": obj.theme_stack, "buffer": obj.buffer, "buffer_index": obj.buffer_index}
+        return create_locals, (kwargs, )
 
-        pickler.save_reduce(unpickle, (args,), obj=obj)
+    _register(ConsoleThreadLocals, pickle_locals)
 
 
-@mainify
-def patch_dill():
-    import dill
+def _patch_exceptions() -> None:
+    # Support chained exceptions
+    from tblib.pickling_support import install
 
-    dill.settings["recurse"] = True
+    install()
 
-    patch_exceptions()
-    patch_pydantic_class_attributes()
-    patch_pydantic_field_serialization()
-    _patch_console_thread_locals()
 
+def patch_pickle() -> None:
+    _patch_pydantic_field_serialization()
+    _patch_lru_cache()
+    _patch_lock()
+    _patch_rlock()
+    _patch_console_thread_locals()
```

Comment:
these 4 the only things i don't really like, but i think we have to do them even though i don't know why (e.g. which object causes this and how do we get rid of it)

Reply:
lru_cache one is a known issue in cloudpickle (there's a link in the code), we use it in our toolkit.

lock/rlock ones happen in our code and in tests, but also generally useful to keep around for user code.

console_thread_locals one pops out in our code sometimes and it is a bit flaky to appear, but again good to have around (we actually patch it in dill in main right now too).

I mentioned this before, but ideally we'll introduce some context/hooks/registry to be able to add user reducers into our serialization context. Obviously they could just copyreg, but keeping stuff more isolated is safer.